SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp

```diff
@@ -28,169 +30,207 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
+				ComponentToClassInfoMap.Emplace(ComponentId, ActorSubobjectInfo);
+				ComponentToOffsetMap.Add(ComponentId, Offset);
+				ComponentToCategoryMap.Add(ComponentId, ESchemaComponentType(Type));
 			}
 		});
 
-		Info.Class = Class;
+		Info.SubobjectInfo.Add(Offset, ActorSubobjectInfo);
+	}
+}
 
-		ClassInfoMap.Emplace(Class, Info);
+UClass* USpatialTypebindingManager::LoadClassForComponent(Worker_ComponentId ComponentId)
+{
+	// Only try to load classes for generated components
+	if (ComponentId < SpatialConstants::STARTING_GENERATED_COMPONENT_ID)
+	{
+		return nullptr;
 	}
 
-	for (UClass* Class : SupportedClasses)
+	for (auto& ObjectDataPair : SchemaDatabase->ClassPathToSchema)
 	{
-		for (auto& SubobjectDataPair : SchemaDatabase->ClassPathToSchema[Class->GetPathName()].SubobjectData)
+		for (int32 Type = SCHEMA_Begin; Type < SCHEMA_Count; Type++)
 		{
-			int32 Offset = SubobjectDataPair.Key;
-			FSubobjectSchemaData SubobjectSchemaData = SubobjectDataPair.Value;
-
-			FClassInfo* ActorInfo = FindClassInfoByClass(Class);
-
-			FSoftClassPath SubobjectClassPath(SubobjectSchemaData.ClassPath);
-			UClass* SubobjectClass = SubobjectClassPath.TryLoadClass<UObject>();
-			if (SubobjectClass == nullptr)
+			const Worker_ComponentId ObjectComponentId = ObjectDataPair.Value.SchemaComponents[Type];
+			if (ComponentId == ObjectComponentId)
 			{
-				continue;
-			}
-
-			FClassInfo* SubobjectInfoPtr = FindClassInfoByClass(SubobjectClass);
-			if (SubobjectInfoPtr == nullptr)
-			{
-				continue;
+				FSoftClassPath SoftClassPath(ObjectDataPair.Key);
+				UClass* Class = SoftClassPath.TryLoadClass<UObject>();
+				if (Class == nullptr)
+				{
+					UE_LOG(LogSpatialTypebindingManager, Warning, TEXT("Failed to load class at path %s which is needed for component %u"),
+						*ObjectDataPair.Key, ComponentId);
+				}
+				AddTypebindingsForClass(Class);
+				return Class;
 			}
+		}
 
-			// Make a copy of the already made FClassInfo for this specific subobject
-			FClassInfo SubobjectInfo = *SubobjectInfoPtr;
-
-			SubobjectInfo.SubobjectName = SubobjectSchemaData.Name;
-
-			ForAllSchemaComponentTypes([&](ESchemaComponentType Type)
+		for (auto& SubobjectDataPair : ObjectDataPair.Value.SubobjectData)
+		{
+			for (int32 Type = SCHEMA_Begin; Type < SCHEMA_Count; Type++)
 			{
-				Worker_ComponentId ComponentId = SubobjectSchemaData.SchemaComponents[Type];
-				if (ComponentId != 0)
+				const Worker_ComponentId SubobjectComponentId = SubobjectDataPair.Value.SchemaComponents[Type];
+				if (ComponentId == SubobjectComponentId)
 				{
-					SubobjectInfo.SchemaComponents[Type] = ComponentId;
-					ComponentToClassMap.Add(ComponentId, SubobjectClass);
-					ComponentToOffsetMap.Add(ComponentId, Offset);
-					ComponentToCategoryMap.Add(ComponentId, (ESchemaComponentType)Type);
+					FSoftClassPath SoftClassPath(SubobjectDataPair.Value.ClassPath);
+					UClass* Class = SoftClassPath.TryLoadClass<UObject>();
```

Comment:
Do we need to explicitly load the subobject's class?
Supposedly loading the actor class should also load all its components' classes.
@m-samiec @Vatyx any thoughts?

Reply:
In fact, this should never need to load the class, only check that it is already loaded and get the class pointer, which you can do with `SoftClassPath.ResolveClass()`. If it returns null, that would indicate a problem with our logic, so we should just `check` the result of `ResolveClass()`.
The reason for this is that we should have already loaded the class and all of the subobjects' classes when creating the actor, which happens before we go into this.

Reply:
I agree the sub-object class doesn't need to be loaded with the current implementation as it would be loaded as part of the call to `AddTypebindingsForClass(ActorClass)` a few lines below.

Regarding your second point, I'm not familiar with exactly how the actors are created, but without assuming anything it seems possible to me that the sub-object class is not already loaded at this point and will in fact be loaded in a call to  `AddTypebindingsForClass`.

Reply:
I might be wrong, but my current understanding is that we don't even need to load the class in `AddTypebindingsForClass`, only try to resolve it. The reasoning is that the entity would have to have been checked out before we could receive updates or commands on it, which means that we would need to spawn the corresponding actor first, which means that the actor class as well as the subobject classes would be already loaded.

Reply:
Tested this and it seems to work ok.